File: tests/test_command_init.py
# pyright: standard

import json
from pathlib import Path

from typer.testing import CliRunner

from aico.consts import SESSION_FILE_NAME
from aico.main import app

runner = CliRunner()


def test_init_creates_session_file_in_empty_dir(tmp_path: Path) -> None:
    # GIVEN a directory without a session file
    # We use pytest's tmp_path fixture and run the command within that isolated directory.
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        # WHEN `aico init` is run
        result = runner.invoke(app, ["init"])

        # THEN the command succeeds and creates the session file
        assert result.exit_code == 0
        pointer_file = Path(td) / SESSION_FILE_NAME
        assert pointer_file.is_file()
        assert f"Initialized session file: {pointer_file.name}" in result.stdout

        # AND the pointer file has the correct format and points to a view
        pointer_data = json.loads(pointer_file.read_text())
        assert pointer_data["type"] == "aico_session_pointer_v1"
        view_path = Path(td) / pointer_data["path"]
        assert view_path.is_file()

        # AND the view file contains the default model
        view_data = json.loads(view_path.read_text())
        assert view_data["model"] == "openrouter/google/gemini-3-pro-preview"


def test_init_fails_if_session_already_exists(tmp_path: Path) -> None:
    # GIVEN a directory with an existing session file
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        expected_path = Path(td) / SESSION_FILE_NAME
        expected_path.touch()

        # WHEN `aico init` is run again
        result = runner.invoke(app, ["init"])

        # THEN the command fails with an appropriate error message
        assert result.exit_code == 1

        assert f"Error: Session file '{expected_path}' already exists in this directory." in result.stderr


def test_init_creates_gitignore(tmp_path: Path) -> None:
    # GIVEN an empty project directory
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        # WHEN `aico init` is run
        result = runner.invoke(app, ["init", "--model", "test-model"])
        assert result.exit_code == 0

        # THEN the .aico/.gitignore file is created with the correct content
        gitignore_path = Path(td) / ".aico" / ".gitignore"
        assert gitignore_path.is_file()
        assert gitignore_path.read_text() == "*\n!addons/\n!.gitignore\n"


def test_init_creates_secure_directories(tmp_path: Path) -> None:
    """Tests that init creates .aico directories with secure 0700 permissions."""

    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        result = runner.invoke(app, ["init", "--model", "test-model"])
        assert result.exit_code == 0

        aico_dir = Path(td) / ".aico"
        history_dir = aico_dir / "history"
        sessions_dir = aico_dir / "sessions"

        for dir_path in [aico_dir, history_dir, sessions_dir]:
            assert dir_path.is_dir()
            mode = dir_path.stat().st_mode & 0o777
            assert mode == 0o700, f"Expected 0700 for {dir_path}, got {oct(mode)}"

File: tests/test_command_init.rs
use assert_cmd::cargo;
use predicates::prelude::*;
use std::fs;

const SESSION_FILE_NAME: &str = ".ai_session.json";

#[test]
fn test_init_creates_session_file_in_empty_dir() {
    let temp = tempfile::tempdir().unwrap();

    let mut cmd = cargo::cargo_bin_cmd!("aico");
    cmd.current_dir(&temp)
        .arg("init")
        .assert()
        .success()
        .stdout(predicate::str::contains(format!(
            "Initialized session file: {}",
            SESSION_FILE_NAME
        )));

    // Check pointer - Verify exact JSON structure for cross-compatibility
    let pointer_path = temp.path().join(SESSION_FILE_NAME);
    assert!(pointer_path.exists());
    let pointer_content = fs::read_to_string(&pointer_path).unwrap();
    let expected_pointer_json =
        r#"{"type":"aico_session_pointer_v1","path":".aico/sessions/main.json"}"#;
    assert_eq!(pointer_content, expected_pointer_json);

    // Check view
    let view_path = temp.path().join(".aico/sessions/main.json");
    assert!(view_path.exists());
    let view_content = fs::read_to_string(view_path).unwrap();
    assert!(view_content.contains("openrouter/google/gemini-3-pro-preview"));
}

#[test]
fn test_init_fails_if_session_already_exists() {
    let temp = tempfile::tempdir().unwrap();
    let session_file = temp.path().join(SESSION_FILE_NAME);
    fs::write(&session_file, "existing").unwrap();

    let mut cmd = cargo::cargo_bin_cmd!("aico");
    cmd.current_dir(&temp)
        .arg("init")
        .assert()
        .failure() // Should exit 1
        .stderr(predicate::str::contains("already exists"));
}

#[test]
fn test_init_creates_secure_directories() {
    let temp = tempfile::tempdir().unwrap();

    let mut cmd = cargo::cargo_bin_cmd!("aico");
    cmd.current_dir(&temp).arg("init").assert().success();

    #[cfg(unix)]
    {
        use std::os::unix::fs::MetadataExt;
        let mode = fs::metadata(temp.path().join(".aico")).unwrap().mode();
        assert_eq!(mode & 0o777, 0o700);
    }
}

#[test]
fn test_init_creates_gitignore() {
    let temp = tempfile::tempdir().unwrap();

    assert_cmd::cargo::cargo_bin_cmd!("aico")
        .current_dir(&temp)
        .arg("init")
        .assert()
        .success();

    let gitignore_path = temp.path().join(".aico/.gitignore");
    assert!(gitignore_path.exists());
    let content = fs::read_to_string(gitignore_path).unwrap();
    assert_eq!(content, "*\n!addons/\n!.gitignore\n");
}
